bin/src/notifications.rs

```diff
@@ -0,0 +1,152 @@
+    failed_since_last_success: u32,
+}
+
+pub struct NotificationService {
+    channel_name: Option<String>,
+    tx: mpsc::Sender<Notification>,
+    rx: mpsc::Receiver<Notification>,
+}
+
+impl NotificationService {
+    pub fn new(channel_name: Option<String>) -> Self {
+        let (tx, rx) = mpsc::channel(10);
+        Self {
+            channel_name,
+            tx,
+            rx,
+        }
+    }
+
+    pub fn new_handler(&self) -> mpsc::Sender<Notification> {
+        return self.tx.clone();
+    }
+}
+
+pub fn run_notification_worker(
+    task_tracker: &TaskTracker,
+    mut notification_service: NotificationService,
+) {
+    task_tracker.spawn(async move {
+        let channel_name = notification_service.channel_name.unwrap_or("".into());
+
+        if channel_name != String::from("") {
+            let mut notification_statuses = HashMap::new();
+            let http_client = reqwest::Client::new();
+
+            loop {
+                // Get any notifications since the last time we checked
+                loop {
+                    match notification_service.rx.try_recv() {
+                        Ok(notification) => {
+                            let status = notification_statuses
+                                .entry(notification.message_type)
+                                .or_insert_with(|| NotificationStatus {
+                                    message: "".to_string(),
+                                    needs_sending: true,
+                                    last_sent: None,
+                                    last_attempt: None,
+                                    failed_since_last_success: 0,
+                                });
+                            // Ignore if we're in the debounce period
+                            if let Some(debounce) = notification.debounce {
+                                if let Some(last_sent) = status.last_sent {
+                                    if last_sent.elapsed() < debounce {
+                                        continue;
+                                    }
+                                }
+                            }
+                            status.message = notification.message;
+                            status.needs_sending = true;
+                        }
+                        Err(TryRecvError::Empty) => {
+                            break;
+                        }
+                        Err(TryRecvError::Disconnected) => {
+                            return;
+                        }
+                    }
+                }
+
+                // Attempt to send pending notifications
+                for notification in notification_statuses.values_mut() {
+                    if !notification.needs_sending {
+                        continue;
+                    }
+
+                    // Backoff retries, up to a maximum of 256 seconds.
+                    if let Some(last_attempt) = notification.last_attempt {
```

Comment:
what would make one notification `message_type` fail to send but another one succeed? size/shape? asking because a global retry would be a bit simpler

Reply:
I was thinking this way would be simpler, but thinking about it again you might be right. I can try switching it and seeing how it goes.

Reply:
if you want to drop all notifications but the last one per `message_type` then i think you won't save much actually. i was thinking you could enter into retry mode and just not consume from the mpsc while you are retrying. but then you might end up backpressuring into the sender or dropping notifications.